protocol/consistencychecks.go

```diff
@@ -215,6 +256,45 @@ func (cc *ConsistencyChecks) verifyKeyLookup(msg *Response,
 	return CheckPassed
 }
 
+func (cc *ConsistencyChecks) verifyMonitoring(msg *Response,
+	uname string, key []byte) error {
+	if msg.Error != ReqSuccess {
+		return ErrMalformedDirectoryMessage
+	}
+	dfs := msg.DirectoryResponse.(*DirectoryProofs)
+
+	str0 := dfs.STR[0]
+	ap0 := dfs.AP[0]
+	regEp, ok := cc.RegEpoch[uname]
+
+	switch {
+	case !ok && str0.Epoch == 0 && ap0.ProofType() == m.ProofOfAbsence: /* prior history verification */
+	case ok && str0.Epoch == regEp && ap0.ProofType() == m.ProofOfAbsence: /* registration epoch */
+	case ok && str0.Epoch >= regEp+1 && ap0.ProofType() == m.ProofOfInclusion: /* after registration */
+	default:
+		return CheckBadAuthPath
+	}
+	if err := verifyAuthPath(uname, key, ap0, str0); err != nil {
+		return err
+	}
+
+	for i := 1; i < len(dfs.STR); i++ {
+		str := dfs.STR[i]
+		ap := dfs.AP[i]
```

Comment:
Sure. I changed a bit in `HandleResponse` to explicitly validate each response type instead of calling `msg.validate()`.